feat: Add NoteBinary* functions for encoding and decoding.

haydenroche5 opened the pull request:
And make the existing NoteBinary functions use these instead of the cobs* ones.

zfields commented:
Let's put a pin in the this until we (you, Mat and I) get a chance to talk through the long-term ramifications of the API.

haydenroche5 commented on n_helpers.c:
```diff
@@ -84,6 +84,45 @@ NOTE_C_STATIC void setTime(JTIME seconds);
+{
+    uint32_t decLen = cobsDecode((uint8_t *)inBuf, inLen, BINARY_EOP, outBuf);
+
+    if (outLen != NULL) {
```
In `NoteBinaryTransmit`, we decode the user's input buffer on errors. In those cases, we have no use for the output length, so I deliberately made this function permit `outLen == NULL`.

m-mcgowan replied:
Yes, if we can ascertain the lower bound on the required length for the decoded data, that's preferable to just trying and hoping it doesn't crash. 

Is there a better name for `cobsGuaranteedFit()`? To me, it's not immediately obvious this is talking about decoding space required, and still wasn't all that clear after I read the doc comments.

zfields replied:
I agree, but that is Ray's naming convention and it is taken directly from the Notecard firmware. My goal is to keep these files identical.

Changing the name is a reasonable request, but it should be done outside of this PR.

haydenroche5 commented on n_helpers.c:
```diff
@@ -348,25 +393,27 @@ const char * NoteBinaryTransmit(uint8_t * data, size_t dataLen, size_t bufLen, s
         return ERRSTR("buffer size exceeds available memory", c_mem);
     }
 
-    // Confirm buffer is large enough to in place encode as COBS
```
`NoteBinaryEncode` does this check for us now, so I removed it.

zfields commented on n_helpers.c:
```diff
@@ -348,25 +393,27 @@ const char * NoteBinaryTransmit(uint8_t * data, size_t dataLen, size_t bufLen, s
+    // packet. When NoteBinaryEncode returns, outLen will hold the encoded
+    // length.
+    uint32_t outLen = bufLen - 1;
+    err = NoteBinaryEncode(data + dataShift, dataLen, data, &outLen);
```
This is a reasonable use of intra-interface calls. I see `NoteBinaryTransmit` and `NoteBinaryReceive` as second order functions, and the diff illustrates the value provided.

m-mcgowan replied:
Agreed, andI feel the same applies to all the cases mentioned.

zfields commented on n_helpers.c:
```diff
@@ -389,33 +436,33 @@ const char * NoteBinaryTransmit(uint8_t * data, size_t dataLen, size_t bufLen, s
                 // decoding it. The caller is then able to retry transmission
                 // with their original pointer to this buffer.
-                cobsDecode(data, encLen, '\n', data);
+                NoteBinaryDecode(data, outLen, data, NULL);
```
Don't love the `NULL` parameter usage here, should be `bufLen`, no?

haydenroche5 replied:
https://github.com/blues/note-c/pull/90#discussion_r1304978894

zfields commented on n_helpers.c:
```diff
@@ -376,7 +423,7 @@ const char * NoteBinaryTransmit(uint8_t * data, size_t dataLen, size_t bufLen, s
         J *req = NoteNewRequest("card.binary.put");
         if (req) {
-            JAddIntToObject(req, "cobs", encLen);
+            JAddIntToObject(req, "cobs", outLen);
```
I don't understand this rename.

`encLen` (encoded length) seems more expressive in the context of this function. If it was renamed to match the parameters of `NoteBinaryDecode`, then should it not be `inLen`?

haydenroche5 replied:
It's an in/out parameter now. On input, it's the length of the output buffer, so encLen didn't feel like a great name anymore. I felt outLen kind of captured the semantics of both input and output: on input, it's the length of the output buffer, and on output, it's the length of the actual data in the output buffer.

m-mcgowan commented on n_helpers.c:
```diff
@@ -84,6 +84,45 @@ NOTE_C_STATIC void setTime(JTIME seconds);
+const char *NoteBinaryEncode(const uint8_t *inBuf, uint32_t inLen,
+                             uint8_t *outBuf, uint32_t *outLen)
+{
+    if (*outLen < NoteBinaryEncodedMaxLength(inLen)) {
```
It's just to optimize performance.  The first check is O(1) and is a quick check. The second check is O(N).

zfields replied:
Cool, that sounds great. I completely overlooked that.

haydenroche5 commented on n_helpers.c:
```diff
@@ -84,6 +84,45 @@ NOTE_C_STATIC void setTime(JTIME seconds);
+                             uint8_t *outBuf, uint32_t *outLen)
+{
+    if (*outLen < NoteBinaryEncodedMaxLength(inLen)) {
+        if (*outLen < NoteBinaryEncodedLength(inBuf, inLen)) {
```
I'm fine either way, but I'm curious what an example of this problem looks like. Not sure I've seen it myself.

zfields replied:
Mat did a really good job of breaking down the "levels" of the API.

> 1.  NoteBinaryTransmit/Receive and the helpers for required Transmit/Receive buffer sizes.
> 1.  NoteBinaryEncode/Decode/[Length]
> 1.  NoteBinaryMaxEncodedLength

When you co-mingle in this fashion, you are assuming that all engineers that touch this code will be able to make the same break down in their head.

haydenroche5 commented on n_helpers.c:
```diff
@@ -169,8 +208,12 @@ const char * NoteBinaryReceive(uint8_t * buffer, size_t bufLen,
+    uint32_t decLen = bufLen;
     // Decode it in-place, which is safe because decoding shrinks
-    const size_t decLen = cobsDecode(buffer, bufLen, '\n', buffer);
+    err = NoteBinaryDecode(buffer, bufLen, buffer, &decLen);
```
- If we change the codec to something other than COBS, then we don't have to change this code again. It just works.
- We get test coverage of the new functions "for free."
- We are using the same functions internally that our users will be using externally, which makes us more confident that they actually work/are bug-free.

I'm not opposed to switching over to the cobs* functions, but I'm not seeing the benefit to doing that, and like I commented above, I'm not sure what an example of the problem you're talking about looks like in practice.

zfields replied:
Personally, I see as two levels of API. The level of API that facilitate the direct usage of the Notecard API (levels 2 and 3), and the level  that completely abstracts it (level 1).

I really have to squint to see all three levels and not just two, but I do understand how you're making the distinction.

I think you are being optimistic that we wouldn't need to change all the implementations, but I don't have an overly strong opinion here.

haydenroche5 commented:
I think I addressed all the review comments. @zfields and I discussed in a Slack huddle and settled on using the cobs* functions internally for the new functions. For NoteBinaryReceive/Transmit, we use the new functions. Latest push also adds Doxygen comments.

zfields commented on n_helpers.c:
```diff
@@ -84,6 +84,89 @@ NOTE_C_STATIC void setTime(JTIME seconds);
+const char *NoteBinaryEncode(const uint8_t *inBuf, uint32_t inLen,
+                             uint8_t *outBuf, uint32_t *outLen)
+{
+    if (*outLen < cobsEncodedMaxLength(inLen)) {
```
It's dangerous to dereference a pointer (especially on a public API) without first checking it for `NULL`.

haydenroche5 replied:
I agree. I just don't see us coding defensively like that in note-c, historically-speaking, but I think it would be good to start. I'll add checks.

zfields commented on n_helpers.c:
```diff
@@ -348,25 +437,27 @@ const char * NoteBinaryTransmit(uint8_t * data, size_t dataLen, size_t bufLen, s
+    // one byte of space we need to save for a newline to mark the end of the
+    // packet. When NoteBinaryEncode returns, outLen will hold the encoded
+    // length.
+    uint32_t outLen = bufLen - 1;
```
> // outLen holds the buffer size available for encoding.

This comment is only necessary, because you have named the variable to match the parameter (context) of the function you are calling into, as opposed to the context of the current function in which it is being used.

haydenroche5 replied:
Oh yeah I meant to change this back to `encLen` per our conversation.
